lib/client/reexec.go

```diff
@@ -0,0 +1,165 @@
+		return trace.Wrap(err)
+	}
+	return runForkAuthenticateChild(ctx, cmd)
+}
+
+func buildForkAuthenticateCommand(params ForkAuthenticateParams) (*forkAuthCmd, error) {
+	executable, err := getExecutable()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	cmd := exec.Command(executable)
+	// Set up disown signal.
+	signalR, signalW, err := os.Pipe()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	killR, killW, err := os.Pipe()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	signalFd, killFd := configureReexecForOS(cmd, signalW, killR)
+
+	cmd.Args = append(cmd.Args, params.GetArgs(signalFd, killFd)...)
+	cmd.Args[0] = os.Args[0]
+	cmd.Stdin = params.Stdin
+	cmd.Stdout = params.Stdout
+	cmd.Stderr = params.Stderr
+
+	return &forkAuthCmd{
+		Cmd: cmd,
+		// Keep all pipes around to prevent them from being garbage collected.
+		signalR: signalR,
+		signalW: signalW,
+		killR:   killR,
+		killW:   killW,
+	}, nil
+}
+
+func (cmd *forkAuthCmd) killProcess() error {
+	// It's ok if the child finished before Kill() ran.
+	if err := cmd.Process.Kill(); err != nil && !strings.Contains(err.Error(), "os: process already released") {
+		return trace.Wrap(err)
+	}
+	return trace.Wrap(cmd.Process.Release())
+}
+
+func runForkAuthenticateChild(ctx context.Context, cmd *forkAuthCmd) error {
+	defer func() {
+		// If the child is still listening, kill it. If the child successfully
+		// disowned, this will do nothing.
+		cmd.killW.Write([]byte{0x00})
+		cmd.killW.Close()
+		cmd.signalR.Close()
+	}()
+
+	disownReady := make(chan error, 1)
+	go func() {
+		// The child process will write to the pipe when it has authenticated
+		// and is ready to be disowned.
+		n, err := cmd.signalR.Read(make([]byte, 1))
+		if n > 0 {
+			disownReady <- nil
+		} else if err == nil {
+			// this should be impossible according to the io.Reader contract
+			disownReady <- io.ErrUnexpectedEOF
+		} else {
+			disownReady <- err
+		}
+	}()
+
+	if err := cmd.Start(); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := cmd.signalW.Close(); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := cmd.killR.Close(); err != nil {
+		return trace.Wrap(err)
+	}
+	select {
+	case err := <-disownReady:
+		if err == nil {
+			return trace.Wrap(cmd.Process.Release())
+		}
+		// Error was likely caused by the child process exiting. Wait for Wait() to
+		// return the exit status if possible.
+		childFinished := make(chan error, 1)
+		go func() {
+			childFinished <- cmd.Wait()
+		}()
+		select {
+		case err := <-childFinished:
+			return trace.Wrap(err)
+		case <-time.After(5 * time.Second):
+			if err := cmd.killProcess(); err != nil {
+				return trace.Wrap(err)
+			}
+			// Give Wait() one last chance to finish.
+			select {
+			case err := <-childFinished:
+				return trace.Wrap(err)
+			case <-time.After(3 * time.Second):
+				return trace.Errorf("timed out waiting for child process to finish")
+			}
+		}
+
+	case <-ctx.Done():
+		return trace.Wrap(cmd.killProcess())
```

Comment:
This is the only occurrence of ctx in this - what happens if the context actually expires? Won't it result in a zombie process, if the process doesn't cooperate?